ypathgen/pathgen.go

```diff
@@ -291,18 +303,63 @@ func (cg *GenConfig) GeneratePathCode(yangFiles, includePaths []string) (*Genera
 		if cg.GenerateWildcardPaths {
 			listBuilderKeyThreshold = cg.ListBuilderKeyThreshold
 		}
-		structSnippet, es := generateDirectorySnippet(directory, directories, schemaStructPkgAccessor, cg.PathStructSuffix, listBuilderKeyThreshold, cg.GenerateWildcardPaths, cg.SimplifyWildcardPaths)
+		fakeRootPkgName := defaultFakeRootPkgName
+		if cg.FakeRootPackageName != "" {
+			fakeRootPkgName = cg.FakeRootPackageName
+		}
+
+		structSnippet, es := generateDirectorySnippet(directory, directories, schemaStructPkgAccessor, cg.PathStructSuffix, listBuilderKeyThreshold, cg.GenerateWildcardPaths, cg.SimplifyWildcardPaths, cg.SplitByModule, cg.TrimOCPackage, cg.PackageName, fakeRootPkgName)
 		if es != nil {
 			errs = util.AppendErrs(errs, es)
 		}
-		structSnippets = append(structSnippets, structSnippet)
+		structSnippets = append(structSnippets, structSnippet...)
+	}
+
+	// Group snippets by package and compute their deps.
+	packages := map[string]*GeneratedPathCode{}
+	for _, snippet := range structSnippets {
+		if _, ok := packages[snippet.Package]; !ok {
+			packages[snippet.Package] = &GeneratedPathCode{
+				Deps: make(map[string]bool),
+			}
+		}
+		packages[snippet.Package].Structs = append(packages[snippet.Package].Structs, snippet)
+		for _, d := range snippet.Deps {
+			packages[snippet.Package].Deps[d] = true
+		}
+	}
+	for name, p := range packages {
+		err := writeHeader(yangFiles, includePaths, name, cg, p)
+		util.AppendErr(errs, err)
 	}
 	genCode.Structs = structSnippets
 
 	if len(errs) == 0 {
 		errs = nil
 	}
-	return genCode, nodeDataMap, errs
+	return packages, nodeDataMap, errs
+}
+
+// packageNameReplacePattern matches all characters allowed in yang modules, but not go packages.
+var packageNameReplacePattern = regexp.MustCompile("[._-]")
+
+// goPackageName returns the go package to use when generating code for the input Directory.
+// If splitByModule is false, the pkgName is always returned. Otherwise,
+// a transformed version of the module that the directory belongs to is returned.
+func goPackageName(dir *ygen.Directory, splitByModule, trimOCPkg bool, pkgName, fakeRootPkgName string) string {
```

Comment:
So today I've also received a complaint about the generated size of the GoStructs, which was an option you mentioned but we didn't investigate at the time due to the extra complexity of GoStructs. I took a look today and I think it's possible to also split GoStructs in the future while still having a top-level package that users can refer to. Since this is the case, maybe the package names for the PathStructs should have "path" appended to them to keep them separate from GoStructs. I like the idea of keeping them separate to further reduce the chances of a package that's too big.

Reply:
So essentially the idea is to split the GoStructs, and then use type aliasing at the top-level package to give the illusion that they're part of the same package.

Reply:
added _path suffix